### ISSUE TITLE
i18n: Support generating pot files

### ISSUE THREAD

**Issue by grappler** [ahmadawais/clai-old#24]
In Gutenberg you can run `npm run gettext-strings` which runs `cross-env BABEL_ENV=gettext webpack` to create a POT file. We should look into what it takes to have a POT file created for a plugin.

**Comment by ahmadawais** [comment]
Definitely! Would you like to contribute that? I'll welcome that change. You'd add that to the scripts folder in `cgb-scripts` packages inside the `packages` folder of this repo. It should run what is needed to build a pot file.

**Comment by grappler** [comment]
After looking at it in more detail I would like to find a better way then copying https://github.com/WordPress/gutenberg/blob/master/i18n/babel-plugin.js to the plugin and using that.

**Comment by grappler** [comment]
This is where the plugin is registered https://github.com/WordPress/gutenberg/blob/d907d35/.babelrc#L37-L43

**Comment by ahmadawais** [comment]
@grappler Yes, I can see that. We can add this to the feature list of version 2.0.0 and you can work on it if you'd like to contribute.

We can create a new package for this in the `packages` dir and then improve `babel-preset-cgb`. What do you say?

**Comment by ahmadawais** [comment]
@grappler Any thoughts?

**Comment by Viper007Bond** [comment]
Isn't using POT files kinda deprecated now? https://make.wordpress.org/meta/handbook/documentation/translations/

**Comment by ahmadawais** [comment]
@grappler Any updates on this one?

**Comment by grappler** [comment]
This has been partly started with https://github.com/ahmadawais/create-guten-block/commit/6c92f4af8f3fe156c4ad4770255edfa63f1facc2

**Comment by Viper007Bond** [comment]
You're right. Even Gutenberg itself generates a POT file: https://github.com/WordPress/gutenberg/tree/master/i18n

**Comment by ahmadawais** [comment]
@grappler I'd be happy to see you lead this feature of create-guten-block? I can offer help where I can. @Viper007Bond is right. We can mimic Gutenberg's `.POT` file generation and add it here.

**Comment by grappler** [comment]
It looks like Gutenberg is changing their `.POT` generation process. https://github.com/WordPress/gutenberg/pull/6007

@ahmadawais What I can't figure out is which `package.json` do I need to add `"@wordpress/babel-plugin-makepot": "1.0.0",`.

**Comment by ahmadawais** [comment]
@grappler Read this https://github.com/ahmadawais/create-guten-block/blob/master/CONTRIBUTING.md

You'd want to do all that in `cgb-scripts` folder and it's package.json. This is a monorepo which means it hosts multiple packages in one repo. The root is `create-guten-block` CLI which has nothing to do with the process, it only creates a folder and install cgb-scripts + package.json file to it and rest is all done by `packages/cgb-scripts`.

**Comment by grappler** [comment]
Here is an example how the `make-pot` can be setup for a single block https://github.com/swissspidy/gutenberg-i18n-block/blob/master/package.json

I am not sure how to add the `babel` task to the `build.js`.

**Comment by ahmadawais** [comment]
@grappler here's the babel config for create-guten-block https://github.com/ahmadawais/create-guten-block/tree/master/packages/babel-preset-cgb

and here is the webpack https://github.com/ahmadawais/create-guten-block/tree/master/packages/cgb-scripts/config

**Comment by maximebj** [comment]
Hope you'll add this feature soon! 

FYI
A friend of mine use pot-to-php and makepot to do this with his tool.
https://github.com/loicblascos/gutenberg-custom-blocks

**Comment by ahmadawais** [comment]
I can add it pretty easily, I am just so short on time nowadays. Also, I am confident @grappler will soon pick it up 👍 and will add this feature for the upcoming release.

**Comment by ahmadawais** [comment]
Also pretty sure WPCLI cannot translate from JavaScript.

**Comment by ahmadawais** [comment]
That's sad! @asharirfan 🤔

**Comment by swissspidy** [comment]
Quick update on https://github.com/wp-cli/i18n-command on https://github.com/swissspidy/gutenberg-i18n-block:

* Work on the WP-CLI command has been completed and it's part of WP-CLI 2.0, which was released yesterday.
* The Gutenberg I18N Block contains different branches with different examples: with or without Babel, with or without WP-CLI
* The WP-CLI command extracts strings from JS and PHP

For plugins hosted on WordPress.org:

Until we're able to update WordPress.org and WordPress core, your best bet is to generate a PHP file containing the extracted strings, just like Gutenberg does it. This is to ensure that WordPress.org can find these strings and users can translate them.

Use `wp.i18n.setLocaleData` and `gutenberg_get_jed_locale_data` to pass the translations to your script.

For private plugins:

Just use WP-CLI to extract all the strings and generate a POT file for you. Load translation files as usual with `load_plugin_textdomain()`.

Use `wp.i18n.setLocaleData` and `gutenberg_get_jed_locale_data` to pass the translations to your script.

**Comment by ahmadawais** [comment]
Thanks for the update @swissspidy 👏

**Comment by ambienthack** [comment]
Came across the problem of block localization for blocks created using create-guten-block. Haven't localized blocks before, so had to go through trial an error. Below is my current workflow.

### Preparation
1.1. create an empty `/languages` folder inside your block plugin

### Generation of translation files
2.1. generate .pot with Loco Translate plugin ( you will have to enable Loco setting _Scan JavaScript files with extensions:_ ). i believe you can use `wp i18n make-pot` instead.
2.2. for every string in .pot, add `#: dist/blocks.build.js:1` source file reference. 
before:
```
#: src/contacts/block.js:17
msgid "Contacts"
msgstr ""

#: src/cover/block.js:17
msgid "Cover"
msgstr ""
``` 
after:
```
#: src/contacts/block.js:17
#: dist/blocks.build.js:1
msgid "Contacts"
msgstr ""

#: src/cover/block.js:17
#: dist/blocks.build.js:1
msgid "Cover"
msgstr ""
```
(see _How .json files are loaded_ for details):
2.3. create .po with Loco Translate or manually
2.4. generate .json translation files with `wp i18n make-json languages --no-purge`

### Loading translation files
3.1. locate generated init.php file inside your plugin
3.2. locate generated function `function your_plugin_name_cgb_block_assets()` in init.php
3.3. inside the function locate `wp_register_script( '[you_plugin_js_handle]', ... )` call and note the handle name
3.4. at the end of `your_plugin_name_cgb_block_assets()` function add `wp_set_script_translations( '[you_plugin_js_handle]', '[your-text-domain]',  plugin_dir_path( dirname(__FILE__) ) . 'languages' );` (replace values inside square brackets with the handle name and the text domain name).

**You are done!** Translations in your block should work now.

### How .json files are loaded 
To register a block, several php function calls should be done (this php code is generated by create-guten-block). One of these calls is `wp_register_script(  '[you_plugin_js_handle]', ...  )`, which registers a script with js code of your block. Note that when you register translations for your block using `wp_set_script_translations( '[you_plugin_js_handle]', ... )`, translations are bound to the same js handle. Also note, that `wp_register_script()` registers js script file with the **compiled** code of your block, located in `dist` folder of your plugin. So, the translations are bound to the **compiled** js file of your block.

Here is an example of translation .json file name: `your-text-domain-ru_RU-a4f783c851c57db1967c3d9a58cfd198.json`. It consists of **three parts**: domain name, locale name and md5 signature of the associated .js file path. So, the third part should be md5('dist/blocks.build.js'), which is a4f783c851c57db1967c3d9a58cfd198. `wp i18n make-json` calculates this automatically and assigns appropriate file names to .json files. It takes file paths from comments inside of  .po files (remember `#: dist/blocks.build.js:1` ?). 

Why doesn't .pot file have the correct js references automatically? Because i18n calls (`__()`, etc) are done inside the **source** js files, located in the `src` folder, so references (created automatically) point to the files inside the `src` folder. That is why step 2.2. is important.

**Comment by swissspidy** [comment]
No need for the manual steps in 2.2 nor using Loco Translate if you just use WP-CLI all the way, and generate the pot file from the build files instead the source folder. Just an FYI :-)

**Comment by ambienthack** [comment]
@swissspidy had no luck with it, i think because compiled files contain module source code inside strings:

```js
(function(module, __webpack_exports__, __webpack_require__) {

"use strict";
eval("/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_classnames__ = __webpack_require__(/*! classnames */ 0);
...
from wp.blocks\n\nregisterBlockType('arbora/cover', {\n\ttitle: __('Cover', 'my-text-domain'), 
...
");
```

am i missing something?

**Comment by swissspidy** [comment]
That shouldn't matter, the important piece is that the function calls like `__('Cover', 'my-text-domain')` aren't mangled. You might need to adjust the webpack/terser config if string extraction can't find the strings from the build files.

**Comment by ambienthack** [comment]
It looks like `wp i18n`  cannot deal with code packed inside eval'ed strings, as it analyzes abstract code tree. 'eval' modules are produced by webpack in create-guten-block dev mode. If you make a production build with `npm build`, produced js files can be parsed with `wp i18n`. Webpack config option of interest is `devtool: 'cheap-eval-source-map'`.
